### ISSUE TITLE
Function List breaks and slows down Notepad++ with certain PHP syntax

### ISSUE THREAD

**Issue by s--g** [notepad-plus-plus/notepad-plus-plus#3514]
### Description of the Issue
There is a bug with the Function List feature where a certain (valid) PHP syntax stops Function List from working and freezes/significantly slows down Notepad++ operations such as saving to disk and switching tabs (I assume any operation that triggers a refresh of the function list).

The bug happens when:

1. There is a PHP class that extends another PHP class with a name **greater than or equal to 20 characters in length**
2. The original class also **implements an interface**

This bug has been there for several versions, and I would say several years.

### Steps to Reproduce the Issue

1.  Create a new PHP file
2. Paste in the following sample code:
```
<?php

// Note: Extended class name is deliberately 20 characters in length
class MyClass extends AAAAAAAAAAAAAAAAAAAA implements AAA
{
    protected function test() { return null; }
}
```
3. Save the file
4. Switch to another tab (ie: another open file in Notepad++) and then switch back to the tab with the sample code
5. Delete 10 characters from the extended class's name (so that it's 10 characters in length)
6. Now try saving the file
7. Now try switching to another tab and then switch back to the tab with the sample code

### Expected Behavior

- At both steps 3 and 6 the file instantly saves
- After steps 3 and 6 the Function List shows a hierarchy which contains the "MyClass" class and the "test" function.
- At steps 4 and 7 switching back to the sample code from another tab instantly switches

### Actual Behavior

- At step 3 there is a delay of about 2+ seconds in saving the file, during which Notepad++ completely freezes
- After step 3 the Function List shows nothing
- At step 4 there is a delay of about 2+ seconds in switching tabs, during which Notepad++ completely freezes
- At step 6 the file saves instantly
- After step 6 the Function List correctly shows a hierarchy which contains the "MyClass" class and the "test" function.
- At step 7 switching back to the sample code from another tab instantly switches

### Debug Information

Notepad++ v7.4.2   (32-bit)
Build time : Jun 18 2017 - 23:34:19
Path : C:\Program Files (x86)\Notepad++\notepad++.exe
Admin mode : OFF
Local Conf mode : OFF
OS : Windows 10 (64-bit)
Plugins : Explorer.dll mimeTools.dll NppConverter.dll NppExport.dll PluginManager.dll

**Comment by MAPJe71** [comment]
Was not able to reproduce the issue.
FYI:
1. comment embedded in class/function header is not supported i.e. remove `// 19 characters in length`;
1. make sure the file ends with a whitespace i.e. space, tab or newline.

**Comment by MetaChuh** [comment]
we are closing this issue due to multiple newer notepad++ versions since it was filed.

please update your notepad++ to the latest version [<kbd>Notepad++ 7.6.6</kbd>](https://notepad-plus-plus.org/download/v7.6.6.html) and migrate or reinstall your plugins.
if your issue still persists, and it is still of relevance to you, please feel free to file a new issue as detailed and reproducible as possible.

thank you for your help and best regards.

-----
```
general notes:

please close your issues as soon as you are satisfied, to help
keeping the notepad++ issue tracker clean.

when creating a new issue, please provide all information as shown at the issue template.
your notepad++ debug info is part of the minimum requirements.
(? > Debug Info... > Copy debug info into clipboard).

please also conduct a search at the issue tracker beforehand, to check if a similar issue
already exists.

for general questions, or if it is not sure, whether your issue is directly related to
the notepad++ source code, please visit us at the notepad++ community forum, and
search if related topics exists. you are welcome to post either at similar
topics, or to create a new topic at Help Wanted or General Discussion.
```
[<p align="center"><kbd> &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; Click here to visit the Notepad++ Community Forum &emsp; &emsp; &emsp; &emsp; &emsp; &emsp;&emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; </kbd></p>](https://notepad-plus-plus.org/community/)

**Comment by Emma-Walker** [comment]
This issue is still not solved in Notepad++ 7.8.6 (32bit).

Notepad++ considerable slows down (3-4 seconds) when saving, opening or switching to a tab containing the code in OP.

**Comment by MAPJe71** [comment]
Still can not reproduce the "hang" issue using N++ 7.8.5 (32bit).

@Emma-Walker your "even simpler code" has different issues and hence does not display anything in the `Funtion List` tree:
1. comments between the class header and the class body is NOT supported;
2. empty classes (i.e. classes not containing function definitions/declarations) will NOT be displayed in the `Funtion List` tree.

**Comment by Emma-Walker** [comment]
Wait, so neither OP's code example nor mine causes your Notepad++ to hang/slow down?

**Comment by MAPJe71** [comment]
Affirmative.

**Comment by Emma-Walker** [comment]
Oh, I think I know why. 

1) You have to save the code as .php
2) You have to Menu -> View -> Function List
